File: solution/977. Squares of a Sorted Array/977_01.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n
        l, r = 0, n - 1
        for i in range(n - 1, -1, -1):
            L = nums[l] * nums[l]
            R = nums[r] * nums[r]
            if L > R:
                res[i] = L
                l += 1
            else:
                res[i] = R
                r -= 1
        return res
    

File: solution/977. Squares of a Sorted Array/977_01.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n);
        
        int l = 0, r = n - 1;
        for (int i = n - 1; i >= 0; --i) {
            int l_square = nums[l] * nums[l];
            int r_square = nums[r] * nums[r];
            
            if (l_square < r_square) {
                ans[i] = r_square;
                --r;
            } else {
                ans[i] = l_square;
                ++l;
            }
        }
        
        return ans;
    }
};
